### DISCUSSION TITLE
Recommendations for error reporting/APM tools like Sentry, Datadog, NewRelic etc that work with the Next 13 app router

### DISCUSSION THREAD

**Discussion by philwolstenholme** [vercel/next.js#47563] [Help]
### Summary

Does anyone have any recommendations/tips for using application performance/error monitoring tools like Sentry, Datadog, NewRelic etc?

Here's what I've found so far:

## Sentry

https://docs.sentry.io/platforms/javascript/guides/nextjs/ mentions support for `app` router is missing, but a public roadmap at https://github.com/getsentry/sentry-javascript/issues/6726 suggests there is _some_ support, with more on the way. Development looks quite active.

## Datadog

https://jake.tl/notes/2021-04-04-nextjs-preload-hack suggests an approach but is not `app` router specific. I'm not sure how it handles Error Boundaries in RSCs, the Next/React monkeypatched Fetch, etc.

https://github.com/vercel/next.js/discussions/16600?sort=new has some ideas but nothing `app` router specific or recent.

## NewRelic

https://github.com/newrelic/newrelic-node-nextjs uses the `-r` (require/preload) approach like in the https://jake.tl link above. It instruments Node but not the browser, but you can do that manually.

https://github.com/newrelic/newrelic-node-nextjs/issues/94 suggests it's tricky to use with Vercel but NewRelic are looking into it, but it's taken a while for the work to be prioritised. 

I couldn't find anything `app` router specific.

**Comment by tocaherge** [comment]
Has anyone found any reasonable way forward in this matter? It's been over a year now.

**Reply by SamRoehrich** [reply]
can you use the instrumentation hook to initialize datadog?

**Reply by tocaherge** [reply]
That's what I'm currently doing, which is fine for getting traces, but I'm more interested in a good error tracking setup for both Server and Client-side errors - which I currently only get by (kind of) by monkeypatching the internal logging á la above linked https://jake.tl/notes/2021-04-04-nextjs-preload-hack. From 2021.

I have found no actual resources giving any concise information about how to get good observability outside of using Vercel and the magic that includes.

I'm also amazed that there is no way to opt out of the obfuscation of errors in Error Boundaries in production. What if I trust my users and want them to see the actual errors...?

**Reply by theogravity** [reply]
You can use the [LogLayer](https://loglayer.dev) logging library to do both [server](https://loglayer.dev/transports/datadog.html) and [client-side](https://loglayer.dev/transports/datadog-browser-logs.html) errors using DataDog. 

https://loglayer.dev/example-integrations/nextjs.html
